### ISSUE TITLE
[dereference.of.nullable] on final field access in closure

### ISSUE THREAD

**Issue by mtf90** [typetools/checker-framework#4873]
```java
import java.util.function.Consumer;

import org.checkerframework.checker.nullness.qual.Nullable;

class CF {

    private final @Nullable Ref ref;

    CF(@Nullable Ref ref) {
        this.ref = ref;
    }

    void fun(CF other) {
        if (this.ref != null && other.ref != null) {
            this.ref.apply(arg -> other.ref.fun());
        }
    }

    static class Ref {
        void fun() {}
        void apply(Consumer<Void> c) { c.accept(null); }
    }
}
```

Running the default `NullnessChecker`

```
checker/bin/javac -processor org.checkerframework.checker.nullness.NullnessChecker CF.java
```

returns the following false-positive

```
CF.java:[15,39] error: [dereference.of.nullable] dereference of possibly-null reference other.ref
```

`other.ref` shouldn't be null at this point because nullness has been checked before and the field is final so it can't be re-assigned by any possible side-effects. I tested against checkerframework 3.17.0 and Java 8.

This issue might be related to #4872 but here, the nullness is checked directly (instead of a result of dataflow) and the access takes place in a closure rather than an inline block.
